rest_log/components/service.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2020 Camptocamp SA (http://www.camptocamp.com)
+# @author Guewen Baconnier <guewen.baconnier@camptocamp.com>
+# @author Simone Orsi <simahawk@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import json
+import traceback
+
+from werkzeug.urls import url_encode, url_join
+
+from odoo import exceptions, registry
+from odoo.http import request
+
+from odoo.addons.component.core import AbstractComponent
+
+
+class RESTServiceDispatchException(Exception):
+
+    rest_json_info = {}
+
+    def __init__(self, message, log_entry_url):
+        super().__init__(message)
+        self.rest_json_info = {"log_entry_url": log_entry_url}
+
+
+class RESTServiceUserErrorException(RESTServiceDispatchException, exceptions.UserError):
+    """User error wrapped exception."""
+
+
+class RESTServiceValidationErrorException(
+    RESTServiceDispatchException, exceptions.ValidationError
+):
+    """Validation error wrapped exception."""
+
+
+class BaseRESTService(AbstractComponent):
+    _inherit = "base.rest.service"
+    # can be overridden to disable logging of requests to DB
+    _log_calls_in_db = True
+
+    def dispatch(self, method_name, *args, params=None):
+        if not self._db_logging_active():
+            return super().dispatch(method_name, *args, params=params)
+        return self._dispatch_with_db_logging(method_name, *args, params=params)
+
+    def _db_logging_active(self):
+        return (
+            request and self._log_calls_in_db and self.env["rest.log"].logging_active()
+        )
+
+    def _dispatch_with_db_logging(self, method_name, *args, params=None):
+        try:
+            result = super().dispatch(method_name, *args, params=params)
```

Comment:
@simahawk would be it possible to use a savepoint here? That would avoid the call to rollback and the creation of a new cursor to log the error. :thinking: 

```python

def _dispatch_with_db_logging(self, method_name, *args, params=None):
    with self._process_exception(self.env, method_name, *args, params),  self.env.cr.savepoint()
        return super().dispatch(method_name, *args, params=params)

@contextmanager
def pro._process_exception(self, env, method_name, *args, params=None):
    try:
        yield
    except Exception e:
        # At this stage the roolback is already done by the savepoint...
        # here we log the exception
        ....
        # and we build a werkzeug response for the exception using the same logic as the one in odoo.addons.base_rest.http -> to be refactored?
        from odoo.http import request
        except_wrapper = request._handle_exception(e)
        headers = except_wrapper.get_body(env)
        body = except_wrapper.get_headers(env)
        return request.make_json_response(body, headers)
```
```

Reply:
seems a good idea, I'll check, tnx!

Reply:
@lmignon I added a TODO for this. I'd prefer to keep as it is right now. Fine for you?

Reply:
OK, even if I'm not a fan of code where we create a new cursor and make explicit call to `rollback` :smirk:

Reply:
@guewen do you remember any special reason to do it this way?

Reply:
No, I suppose it can be replaced by a savepoint. If we can guarantee that nothing has been written to db before the call to `super().dispatch(method_name, *args, params=params)`.

Reply:
> No, I suppose it can be replaced by a savepoint. If we can guarantee that nothing has been written to db before the call to `super().dispatch(method_name, *args, params=params)`.

Can we go ahead as it is now?
I'd prefer to not touch a part that we know is working fine for thousands of requests today.
We have all the time to improve this bit w/out affecting dependencies.
ATM this is blocking (or making more complicated) https://github.com/OCA/wms/pull/156 and https://github.com/OCA/wms/pull/160